chore: add lab for dates (#500)
freeCodeCamp/CurriculumExpansion: frontend-cert/js-projects/dates-lab/user-stories.md

1. You should create a variable named `currentDate` using `let` and assign it the current date and time using the `Date` object.
1. You should create a function named `formatDateMMDDYYYY` that takes a date object as a parameter and returns the date formatted as `MM/DD/YYYY`.
1. Inside the `formatDateMMDDYYYY` function, you should create variables named `month`, `day`, and `year` using `let`.

> 1. The `month` variable should be assigned a String value of the month from the date object using the `getMonth` method and adding 1 to the result, and a `padStart` method to ensure the month is always two digits.

Reviewer: same issue here.
`padStart` and `padEnd` is not taught.

if you think we should teach it, then we can add a video for it.
otherwise, you will need to update this user story

Author: Okay, will update the user stories rather

1. The `day` variable should be assigned a String value of the day from the date object using the `getDate` method.
1. The `year` variable should be the date object's full year using the `getFullYear` method.
1. You should return a template literal string that uses the `month`, `day`, and `year` variables to format the date as `MM/DD/YYYY`.
1. You should create a function named `formatDateLong` that takes a date object as a parameter.

> 1. Inside the `formatDateLong` function, you should create a variable named `options` using `let`.
> 1. Assign `options` an object with a `year` property set to `numeric`, a `month` property set to `long`, and a `day` property set to `numeric`.
> 1. You should return a string that uses the `toLocaleDateString` method on the date object with the `options` object as an argument.

Reviewer: Same thing here.
We should give campers the ability to come up with multiple valid answers like this one.

```js
function formatDateLong(date) {
  return date.toLocaleDateString(undefined, {
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}
```

These user stories can be boiled down to just one that tests if they are returning a string that uses the `toLocaleDateString` method.

```suggestion
1. Your `formatDateLong` function should return a string that uses the `toLocaleDateString` method to format the date in a `Month Day, Year` format. Remember that the `toLocaleDateString` method accepts `locales` and `options` parameters. (Ex `toLocaleDateString(locales, options)`) For the locales, you can pass in `undefined` and for the `options` parameter, you can pass in an object like you learned about in the lecture videos.
```

> 1. You should return a string that uses the `toLocaleDateString` method on the date object with the `options` object as an argument.

Reviewer: The `toLocaleDateString` won't be taught.
So this user story either needs to be changed

Reviewer: If you feel like this method should be taught, then we can add it.
But you will have to let me know 👍

Author: Okay, will update the user stories rather

Author: > If you feel like this method should be taught, then we can add it. But you will have to let me know 👍

I think we can teach the method, getting rid of it will mean I have to rewrite the function in a longer format, but am open to other suggestions too

Reviewer: the lecture video script has been updated to mention toLocaleDateString

1. You should create a variable named `currentDate` and assign it the current date and time using the `Date` object.
1. You should create a function named `formatDateMMDDYYYY` that takes a parameter. You can name this parameter whatever you like.
1. The function should use `getMonth`, `getDate`, and `getFullYear` methods to construct the date string. You will need to add 1 to the month method.
1. You should create a function named `formatDateLong` that takes a parameter. You can name this parameter whatever you like.
1. Your `formatDateLong` function should return a string that uses the `toLocaleDateString` method to format the date in a `Month Day, Year` format. Remember that the `toLocaleDateString` method accepts `locales` and `options` parameters. (Ex `toLocaleDateString(locales, options)`) For the `locales`, you can pass in `undefined` and for the `options` parameter, you can pass in an object like you learned about in the lecture videos.

> 1. You should print the current date and the results of the above functions to the console.

Reviewer 2: Wonder if we should be specific about how the logs should be so we can test against them? E.g: `You should print Formatted Date (MM/DD/YYYY):, formatDateMMDDYYYY(currentDate) to the console.` - maybe not worded exactly like that, but something specific anyway. What do you think @larymak?

Author: From the previous suggestion, it was highlighted the need to encourage learners to structure their output in a way that makes sense to them. as long as the output correctly reflects the current date and the formatted date strings.

Reviewer: I think an easy solution here would be to have campers create a variable called `shortDateFormat` and then have them assign the result to that. Then they just log that variable.

That will make testing way easier and should allow for multiple valid answers like these:

```js
const shortDateFormat = `Formatted Date (MM/DD/YYYY): ${formatDateMMDDYYYY(currentDate)}`;
console.log(formattedDate);
```

```js
const shortDateFormat = 'Formatted Date (MM/DD/YYYY): ' + formatDateMMDDYYYY(currentDate);
```

```js
let someRandomVariableACamperCreated = formatDateMMDDYYYY(currentDate);
let shortDateFormat = 'Formatted Date (MM/DD/YYYY): ' + someRandomVariableACamperCreated;

```

this gives campers options to come up with whatever solution they like as long as it is correct.

same goes longer date format.
have campers create a `longDateFormat` variable, assign the correct result to it, then log it to the console

```js
const longDateFormat = `Formatted Date (Month Day, Year): ${formatDateLong(currentDate)}`;
console.log(longDateFormat);
```

Reviewer: as for the user story, it could be something along these lines

```md
Your `shortDateFormat` variable should hold the string `Formatted Date (MM/DD/YYYY): [shortDateFormat goes here]`. You will need to replace `[shortDateFormat goes here]` with the `shortDateFormat`.
```

1. You should create a variable named `currentDate` and assign it the current date and time using the `Date` object.
1. You should create a function named `formatDateMMDDYYYY` that takes a parameter. You can name this parameter whatever you like.
1. The function should use `getMonth`, `getDate`, and `getFullYear` methods to construct the date string. You will need to add 1 to the month method.

> 1. You should create a variable named `shortDateFormat` that holds the string `Formatted Date (MM/DD/YYYY): [formatted date]`. Replace `[formatted date]` with the result of calling your `formatDateMMDDYYYY` function with the `currentDate` variable.

Reviewer: ```suggestion
1. You should create a variable named `shortDateFormat` and assign it the string `Formatted Date (MM/DD/YYYY): [formatted date]`. Replace `[formatted date]` with the result of calling your `formatDateMMDDYYYY` function with the `currentDate` variable.
```

1. You should create a function named `formatDateLong` that takes a parameter. You can name this parameter whatever you like.

> 1. Your `formatDateLong` function should return a string that uses the `toLocaleDateString` method to format the date in a `Month Day, Year` format. Remember that the `toLocaleDateString` method accepts `locales` and `options` parameters. (Ex `toLocaleDateString(locales, options)`) For the `locales`, you can pass in `undefined` and for the `options` parameter, you can pass in an object like you learned about in the lecture videos.

Reviewer: At second glance, I think we can generalize this even further to allow campers to think more about what to do here.

```suggestion
1. Your `formatDateLong` function should return a string that formats the date in a `Month Day, Year` format. You can review the previous lecture videos on the different ways to format dates.
```

> 1. You should create a variable named `longDateFormat` that holds the string `Formatted Date (Month Day, Year): [formatted date]`. Replace `[formatted date]` with the result of calling your `formatDateLong` function with the `currentDate` variable.

Reviewer: ```suggestion
1. You should create a variable named `longDateFormat` and assign it the string `Formatted Date (Month Day, Year): [formatted date]`. Replace `[formatted date]` with the result of calling your `formatDateLong` function with the `currentDate` variable.
```
